shipyard_blueprints/magnite/shipyard_magnite/utils.py

```diff
@@ -0,0 +1,335 @@
+from shipyard_templates import ShipyardLogger
+from collections import defaultdict
+from shipyard_magnite import errs
+import csv
+from shipyard_magnite.dataclasses.budget_item import BudgetItem
+from shipyard_magnite.dataclasses.budgets import Budgets
+
+from typing import Dict, Any, Union, List
+
+logger = ShipyardLogger.get_logger()
+
+
+def open_csv(file_path: str) -> List[Dict]:
+    """
+    Opens a CSV file and returns the data as a list of dictionaries.
+
+    Args:
+        file_path (str): The path to the CSV file.
+
+    Returns:
+        List[Dict]: The data from the CSV file.
+    """
+    with open(file_path, "r") as file:
+        data = [{k: v for k, v in row.items()} for row in csv.DictReader(file)]
+    return data
+
+
+def group_budget_data_by_campaign(budget_data):
+    """useful because it groups the budget data by campaign_id, which is needed to update a campaigns budget at once"""
+    logger.debug(f"Grouping budget data by campaign...")
+    campaigns = defaultdict(list)
+    for budget_line in budget_data:
+        logger.debug(f"Processing line: {budget_line}")
+        try:
+            campaign_id = budget_line.pop("campaign_id")
+            logger.debug(f"Adding line to campaign {campaign_id}")
+            campaigns[campaign_id].append(budget_line)
+        except KeyError:
+            logger.error(f"Missing 'campaign_id' in line: {budget_line}")
+
+    results = dict(campaigns)
+    logger.debug(f"Grouped campaigns: {results}")
+    return results
+
+
+def transform_data_to_budget_item(raw_data):
+    """
+    Transforms a dictionary of data into a BudgetItem object.
+
+    Args:
+        data (Dict): A dictionary of data.
+
+    Returns:
+        BudgetItem: A BudgetItem object.
+    """
+    return BudgetItem(
+        id=raw_data.get("id") or raw_data.get("budget_item_id"),
+        budget_value=raw_data.get("budget_value"),
+        budget_period=raw_data.get("budget_period"),
+        budget_pacing=raw_data.get("budget_pacing"),
+        budget_metric=raw_data.get("budget_metric"),
+        vast_caching_adjustment=raw_data.get("vast_caching_adjustment"),
+    )
+
+
+def transform_data_to_budgets(raw_data):
+    """
+    Transforms a list of dictionaries into a Budgets object.
+
+    Args:
+        data (List[Dict]): A list of dictionaries.
+
+    Returns:
+        Budgets: A Budgets object.
+    """
+    return Budgets(items=[transform_data_to_budget_item(item) for item in raw_data])
+
+
+# The below functions were used prior to using dataclasses for validation and organization. Keeping since it could prove useful in the future if we want to handle the raw data directly.
+VALID_BUDGET_METRICS = {
+    "net_cost",
+    "gross_cost",
+    "requests",
+    "revenue",
+    "impressions",
+    "completes",
+    "clicks",
+}
+VALID_BUDGET_PERIODS = {"day", "lifetime", "month", "week", "hour"}
+VALID_BUDGET_PACINGS = {"asap", "smooth", "front_loaded", "even"}
```

Comment:
Are these identical to the other "constants" declared in the PR? If so, could they be consolidated somewhere?

Reply:
I am going to opt to just remove the old code completely. Mainly kept it in in case there was a use case where using the dataclasses was too cumbersome